java/ql/lib/ext/threat-grouping.model.yml

```diff
@@ -0,0 +1,56 @@
+extensions:
+
+  # An incomplete set of sources database reads
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sinkModel
+    data:
+      - ["java.io", "OutputStream", True, "write", "(byte[])", "", "Argument[0]", "sql", "manual"]
+      - ["java.io", "OutputStream", True, "write", "(byte[],int,int)", "", "Argument[0]", "sql", "manual"]
+      - ["java.io", "OutputStream", True, "write", "(int)", "", "Argument[0]", "sql", "manual"]
+
+
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sourceModel
+    data:
+      # Create a few
+      - ["java.sql", "PreparedStatement", True, "executeQuery", "()", "", "ReturnValue", "sql", "manual"]
```

Comment:
@atorralba  : Do you have a good recommendation for a framework or parts of the Java standard library where we could model "stored flow sources" (eg. database reads) or is this modelled already somewhere?

Reply:
Currently, we model this in regular CodeQL with the `DatabaseInput` class. Sadly, its only implementation is the return value of `java.sql.ResultSet.getString`. I guess the `java.sql` package is the place to look at in the JDK for this.

Other frameworks and libraries offer interaction with databases: I can think of Hibernate, jOOQ, Spring JDBC, Spring Data, and MyBatis, but there surely are many more. This sounds like something we'd need to spend time researching/modeling if there's interest for it.

Reply:
Thank you for the input!
We are trying to put together a prototype for making the set of sources for query execution configurable (e.g. use more sources of different kinds OR use completely different sources). I know it is a big task to model a library, but for now we just need to make "a couple" of extra sources for a POC. It sounds like `java.sql` is a good place to look (and there are already a couple of examples in this PR - which might be sufficient).

Another question in that matter:
Does it make sense to execute all queries using the same set of sources or do we require this to be query dependent? For C# I am aware of at least a couple of queries (the second order SQL injection queries) that uses "stored" sources (even though not modelled as MaD yet), which means that we can't rely on a "global" set of standard sources to execute all queries against. Does the same hold for Java?

Reply:
Yes, most queries are interested in remote sources, and some want local ones, but there are a few which want completely different sets of sources (e.g. weak cryptography queries use certain strings or lengths (integers) as sources).

Reply:
Thank you!

Reply:
Thanks for the explanation. I'm guessing this means that for these queries with specific sources, we don't want to allow threat models to configure extra sources for them. Zip Slip queries are another one that come to mind.